### PR TITLE
Add a test coverage suite to the build

### PR DESCRIPTION
**Pull request by danielkz**
Did I make the right changes? I'm a little confused about what should I do next with it. 
Also, I tried to use the codecove.io but its documentation is not good enough. 

Please tell me if I missed something 🤔

### CODE REVIEW

**Comment by mmnaseri** [comment]
Let me give you some tutorial to read through.

## How to verify if a build is going to work in TravisCI

To verify if this works, I will replicate what TravisCI would do:

```bash
# Create directory for the code and switch to it
mkdir tmp
cd tmp
# Clone the git repo here
git clone https://github.com/danielkz/summer .
# Fetch all references
git fetch
# Switch to the branch
git checkout feature-007
# Run whatever is under `script`:
cd summer-build
mvn test -B
# If that is successful, run everything under `after_success`:
mvn clean cobertura:cobertura coveralls:report
```

The last step gives me:
```
[ERROR] No plugin found for prefix 'coveralls' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local, central]
```

## Analyzing the cause of the breakage

That is because you have configured the plugin for the `build` module.

### Coding large projects in Java, without Maven

First, let's go over how you would normally code in Java without something like Maven.

Normally, you wouldn't want to write everything yourself. There are libraries that we want to use.

> **What is a library?**
> A library is just a JAR file that contains compiled classes written by somebody else.
> To use a library, you need access to their JAR file, not their source. This is great, because you can
> write code that other people use without giving them your source code (closed source), or release
> the code separately (open source).

So we are using other JAR files.

> **What is a JAR file?**
> A JAR file (Java Archive file) is just a ZIP file that contains the compiled version of your source code.
> If you change the extension of a JAR file from `.jar` to `.zip` you can uncompress it and see its
> content.

When you want to write code using libraries, you are using code that is not in your immediate class path.

> **What is the *classpath*?**
> When you are compiling your Java project, you need to tell the compiler where it can find your
> sources.
>
> You can do this by saying `javac -cp /path/to/project/sources`.
> The class path can contain three kind of things:
> 1. Java source code, which must be under the root of the class path and its folder structure
> must correspond with its package name, i.e. file /path/to/project/sources/x/y/Z.java must have
> `package x.y;`, and it must have a package-visible or *public* class with the same name as
> the file name, i.e. `public class Z {...}` or `class Z {...}`.
> 2. Compiled Java code, with the same restrictions as the one above. Only, they are not human-readable (since they have been compiled) and their extension is `.class` instead of `.java`.
> 3. JAR files. You can specify a JAR file to be the class path, and then the compiler will use the
> files in the JAR file to compile your code. This way, if you are using a class that *you* didn't define
> but was defined in another JAR file, the compiler doesn't complain.

The usual way to write code is that you create a folder for your own code, and another folder (usually called `lib` or `libs`) alongside it which contains the JAR files of all the other libraries that you want to use. Then, you direct your compiler to consult that folder when you want to compile your project.

Your compile code would then look something like:

```bash
javac -cp lib/lib1.jar;lib/lib2.jar;lib/lib3.jar src
```

Since this usually becomes unwieldy, we usually create a shell script to do this for us.

#### The dependency mess

However, think of what would happen if you suddenly decided to upgrade from using somelibrary.jar version 1.0 to somelibrary.jar version 2.0? You will have to manually go and download the JAR, update your lib folder, and update the command you created before.

Worse than that is when you decide to break your own project into pieces. Imagine you have `piece1`, `piece2`, and `piece3`, and each of them depends on the other. Then, to compile `piece3`, you will have to compile `piece1`, updates it as a dependency JAR for `piece2`, include *that* in the class path for `piece3`, and then compile.

This situation can get exponentially worse when your modules grow and your dependencies change.

#### Make is born

That is why people started to use other tools. The first such tool was Make, which could take parameters and could be broken into parts, and you could say which part to execute. So, you could write a "*Makefile*" with a part for compiling `piece1`, another for copying it to the right directory, then another part for the rest, and so on. But this was still fragile, as it wasn't really much of an improvement over having a directory full of shell scripts that you would have to run in the right order.

#### Let the Anthill do your work

Then, people came up with Apache Ant, which would let you do essentially the same thing as Make, but the format was XML, it had a better, less shell-reliant (imagine writing a lot of Make commands for Ubuntu, then migrating to CentOS), and you could even specify that a *target* (a named set of instructions) was dependent on other targets. This way, Ant would first execute those, and then execute your intended target. So, you could create an *all* target which would depend on the others in the right order.

Even with Ant, there was a ton you had to do and configure. Also, extending Ant and adding functionality wasn't easy.

#### All bow before Maven

So, people came up with Apache Maven. Maven has an XML-based syntax ([DSL](https://en.wikipedia.org/wiki/Domain-specific_language)) which is more aware of a regular Java project's model. Actually, POM is an acronym for **P**roject **O**bject **M**odel.

It provides a structured mechanism for building your code, defining your dependencies, and even does the hard work of going out there and downloading the dependencies for you.

Instead of creating a `lib` file for your project, it creates a hidden folder under your home directory called `~/.m2/repository`, and it will put everything there, and calls it your local repository. Anything that is in your repository can be used as a dependency for your module.

To distinguish each project, you need three elements:

* An `artifactId` which is essentially the name of your project/module.
* A `groupId` which is to avoid name collisions with other people (how many people do you think have created projects and called them `summer`?)
* A `version` to let you release and manage multiple versions of your project.

By default, Maven looks for your code under your module directory (which is any directory with a `pom.xml` file in it), and expects it to be under `src/main/java`.

Maven breaks the process of compiling, testing, releasing, deploying, and so on into various targets, which are executing sequentially (and not all of them are executed all the time, and need you to specify them). Maven gives these tasks a name and calls them phases. Some of the most important Maven targets are:

* `clean`: deletes the `target` folder under your module directory. This destroys the results of a previous Maven call.
* `compile`: Compiles everything under `src/main/java` by using your pom.xml to figure out what to pass to the `-cp` option of the Java compiler. The compiled version of your code is then copied to your module directory under `target/classes`. When you run `clean`, you are deleting everything there.
* `test-compile`: Compiles your code using `compile`, then compiles your tests, which it expects to find under `src/test/java`.
* `test`: Runs your tests by first figuring out which test framework you have used (JUnit, TestNG, etc.)
* `package`: Creates a JAR file (though you can tell Maven to create other things, too) and puts it in your `target` directory.
* `install`: Copies the files it created using `package` to the appropriate directory in your local repository. So, for instance if your group ID was `com.gigli.project`, your artifact ID was `cool`, and your version was `1.0.0`, it will create a directory `~/.m2/repository/com/gigili/project/cool/1.0.0`, and then it copies your packages alongside your `pom.xml` into that directory. This allows other modules on your local machine to use that module as a dependency.
* `site`: This can be used to generate a fancy-pants website for your project, though not many people use it. Any time you go on the internet and see a website with a logo on the bottom left saying "Created by Maven", you should know that they have used `mvn site` to generate that site.
* `deploy`: This will upload the packages from your `target` to a remote Maven repository. For instance, if you look at `https://github.com/mmnaseri/spring-data-mock/tree/master/spring-data-mock/pom.xml`, you will see that I have defined a remote repository to deploy that project. That is why if you go to https://mvnrepository.com and search for "spring data mock" you will find it.

#### What comes after Maven?

Many people have come to the realization that many projects nowadays use multiple languages, and while Maven is theoretically able to oversee multiple languages, its syntax isn't as easy as most people would like, and its operation is a little complex. That's why a lot of people use Gradle. I personally don't like Gradle right now, because I haven't had much experience with it.

#### Back to our own project

Now, I think is a good time to briefly go over what each of the modules do:

* `summer-parent` doesn't have any code in it. It only configures the defaults for its children. In and of itself, this module doesn't have any value. Any dependency defined for this project is ignored, because it doesn't produce a JAR file (hence `<packaging>pom</packaging>` which tells Maven that the only thing that matters from this module is the `pom.xml` file). Any build configurations, properties, or dependency versions defined on this module will be inherited by its children.
* `summer-core` is where we will write the core of our application. This module will have source code and will produce a JAR file. That's why we have not changed its packaging (which defaults to `<packaging>jar</packaging>`). It is a child of `summer-parent`, so it will inherit everything from that module.
* `summer-build` is just a convenience module which will also not produce anything. Its only purpose is that we use it to have a single build target that will build other modules in the right order. Imagine what would happen if we changed something in the parent. Then we would need to do `mvn install` in the parent module, to make sure it is available in our local repository, then go to `summer-core` and do the same so that it would pull in our changes from the parent and use it to build a new artifact. Instead, we would rather have one build *reactor* that would build everything in the right order.

That being said, you have defined the plugin on the build reactor, which doesn't do anything. So, when you build, it first builds the parent, asks it to run the coveralls plugin, and the parent complains that it doesn't have the plugin. If you define the plugin on the parent, everything will work, since the core module will inherit the plugin from the parent, and will not complain when being built.

Hope that helps!

File: .travis.yml [hunk 1]
```diff
@@ -5,3 +5,7 @@ jdk:
 script:
   - cd summer-build
   - mvn test -B
+
+after_success:
```
Comment (mmnaseri) [inline .travis.yml]: Also, I think you need to `cd` into the directory before calling `mvn`.

Comment (danielkz) [reply .travis.yml]: That moment when your faith in humanity restored :D

**Comment by mmnaseri** [comment]
Right now you are getting [an error](https://travis-ci.org/danielkz/summer/jobs/317633779#L1530):

```
[ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.3.0:report (default-cli) on project summer-parent: I/O operation failed: No coverage report files found -> [Help 1]
```

**Comment by danielkz** [comment]
@mmnaseri Yeah, I'm [looking into it](https://github.com/trautonen/coveralls-maven-plugin/issues/57), will solve it asap.

**Comment by mmnaseri** [comment]
Actually, you are just fine. Because you are asking it to upload coverage report, when there aren't any tests. So, to fix this, just write a simple test using TestNG. This means you will have to add TestNG as a dependency to your project, and write a very basic test, which doesn't even need to do anything.

Search in mvnrepository.com to find the best version of TestNG.

Then, in your parent module's `pom.xml`, add:

```xml
<dependencyManagement>
    <dependencies>
        ...
    </dependencies>
</dependencyManagement>
```

where you need to replace `...` with the dependency found as I have mentioned.

Then, you need to include the dependency in your `summer-core` module, so that the JAR files for the TestNG library are available to your core module's class path. Then, you can write some tests.

**Comment by danielkz** [comment]
I'm having difficulty in setting up the last part of the issue.
First I have an error when trying to execute `mvn clean cobertura:cobertura coveralls:report` locally which I could not find any solution to it: 

>[ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.3.0:report (default-cli) on project summer-parent: Build error: Either repository token or service with job id must be defined -> [Help 1]


The second one is that I'm not sure if I made the right changes with the TestNG because I still have the [same error](https://github.com/danielkz/summer/pull/9#issuecomment-352245568). Here is a screenshot of my environment:
screen shot 1396-09-27 at 2 47 01 am

**Comment by mmnaseri** [comment]
It seems like your configuration is faulty. Look at [the usage section of their README](https://github.com/trautonen/coveralls-maven-plugin#usage).

File: summer-core/src/test/java/com/danielkz/projects/summer/HelloWorldTest.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.danielkz.projects.summer;
+
+import static org.testng.Assert.*;
+
+public class HelloWorldTest {
+
+    @org.testng.annotations.Test
```
Comment (mmnaseri) [inline summer-core/src/test/java/com/danielkz/projects/summer/HelloWorldTest.java]: Replace the qualified annotation with an import:

```java
import org.testng.annotations.Test;

// later on

@Test
public void ...
```

IntelliJ should help you with this with <kbd>option-enter</kbd>

File: summer-parent/pom.xml [hunk 3]
```diff
@@ -13,6 +13,44 @@
...
+        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
     </properties>
 
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.testng</groupId>
+                <artifactId>testng</artifactId>
+                <version>6.13.1</version>
+                <scope>test</scope>
```
Comment (mmnaseri) [inline summer-parent/pom.xml]: Good job with the scope! 👍

File: summer-core/src/test/java/com/danielkz/projects/summer/HelloWorldTest.java (hunk 2 above)
Comment (danielkz) [reply summer-core/src/test/java/com/danielkz/projects/summer/HelloWorldTest.java]: Cool!

**Comment by mmnaseri** [comment]
It is worth mentioning that coveralls doesn't seem to be configured. Otherwise, I would be seeing third check from coveralls mentioning the code coverage.

**Comment by danielkz** [comment]
Nah, not satisfied at all. Still trying to find my mistake.

**Comment by danielkz** [comment]
OK here is my thoughts on [the problem](https://github.com/danielkz/summer/pull/9#issuecomment-352295190):

In [Set up the Coveralls maven plugin in the build section of the project pom.xml](https://github.com/trautonen/coveralls-maven-plugin#usage) they say:

> If used as a standalone Maven build or with any continuous integration server other than Travis CI, the Coveralls repository token must be provided. 

We use _Travis-CI_ then we don't need to set the repoToken, right ?

On the other hand in [FAQ](https://github.com/trautonen/coveralls-maven-plugin#faq) question number three they say:

> Q: Can I use Java 8 with the plugin?
A: Yes. The Coveralls plugin works fine with Java 8, but the problem is the coverage tools. Currently only tool supporting Java 8 is JaCoCo. You can use JaCoCo in a single module or a multi-module project and let the Coveralls plugin handle the report aggregation. 

Then how did you make your project run with just Cobertura ? 

The answer continues :

> This is not true aggregation though and does not address cross module coverage calculation (see https://github.com/jacoco/jacoco/pull/97)

So I've searched for aggregation and found [this article](https://www.smartics.eu/confluence/display/BLOG/2013/07/22/Using+Aggregate+and+Parent+POMs) and now I have a simple idea of what aggregate is. So Cobertura supports aggregation and that's why we must use it?

About the _[project key](https://github.com/danielkz/summer/pull/9#pullrequestreview-84308204)_ which you've mentioned earlier, did you mean the repo token of coveralls.io ? I did implement that in the travis-ci.org project's setting:
image

Since our project is open source, was that even necessary ? 

I got so paranoid to the point that I've installed a fresh VM in my pc and installed all the files (intellij, jdk8, mvn, git) and then created a new repository with another github account, joined travis, coveralls, and created the whole project from scratch. Unfortunately, I got the same results again. Tested lots of configuration and changes on that repo, but nothing worked. 

Wondering what is it that I'm missing :thinking:

**Comment by mmnaseri** [comment]
> I got so paranoid to the point that I've installed a fresh VM in my pc and installed all the files (intellij, jdk8, mvn, git) and then created a new repository with another github account, joined travis, coveralls, and created the whole project from scratch. Unfortunately, I got the same results again. Tested lots of configuration and changes on that repo, but nothing worked.

Okay, to say that that was excessive is a criminally major understatement. 🤣 

Now, as to the actual issue.

Good catch. I wasn't aware that Cobertura isn't compatible with Java 8. And to answer your question, I didn't get Cobertura to run with Java 8, because that project was written based on Java 7. So, there's that.

We are fine with not having cross module calculation, so let's go with JaCoCo. The one nice thing that I liked about Cobertura is branch coverage (e.g. it checks if you have tests for your IFs for both the `true` and the `false` condition). But, oh well, if we can't have anything with Cobertura, I am happy to have *something* with JaCoCo.

**Comment by danielkz** [comment]
Haha love that red 0% coverage badge :D
If you agree, I think it's time to merge.

Oh man thank you for your [explanation](https://github.com/danielkz/summer/pull/9#issuecomment-352144401). It's perfect and cleared lots of things for me ❤️

**Comment by mmnaseri** [comment]
Super. I think it's time to call this done. Good job 👍

**Pull request merged**